pkgs/development/python-modules/tree-sitter-grammars/default.nix

```diff
@@ -0,0 +1,162 @@
+        }
+
+        static PyMethodDef methods[] = {
+            {"language", _binding_language, METH_NOARGS,
+            "Get the tree-sitter language for this grammar."},
+            {NULL, NULL, 0, NULL}
+        };
+
+        static struct PyModuleDef module = {
+            .m_base = PyModuleDef_HEAD_INIT,
+            .m_name = "_binding",
+            .m_doc = NULL,
+            .m_size = -1,
+            .m_methods = methods
+        };
+
+        PyMODINIT_FUNC PyInit__binding(void) {
+            return PyModule_Create(&module);
+        }
+      '')
+      (writeTextDir "setup.py" ''
+        from platform import system
+        from setuptools import Extension, setup
+
+
+        setup(
+          packages=["${snakeCaseName}"],
+          ext_package="${snakeCaseName}",
+          ext_modules=[
+            Extension(
+              name="_binding",
+              sources=["${snakeCaseName}/binding.c"],
+              extra_objects = ["${grammarDrv}/parser"],
+              extra_compile_args=(
+                ["-std=c11"] if system() != 'Windows' else []
+              ),
+              define_macros=[
+                # Define python limited API for compatibility.
+                # https://docs.python.org/3/c-api/stable.html#c.Py_LIMITED_API
+                ("Py_LIMITED_API", "0x03080000"),
+                ("PY_SSIZE_T_CLEAN", None)
+              ],
+              py_limited_api=True,
+            )
+          ],
+        )
+      '')
+      (writeTextDir "pyproject.toml" ''
+        [build-system]
+        requires = ["setuptools>=42", "wheel"]
+        build-backend = "setuptools.build_meta"
+
+        [project]
+        name="${snakeCaseName}"
+        description = "${langIdent} grammar for tree-sitter"
+        version = "${version}"
+        keywords = ["parsing", "incremental", "python"]
+        classifiers = [
+          "Development Status :: 4 - Beta",
+          "Intended Audience :: Developers",
+          "License :: OSI Approved :: MIT License",
+          "Topic :: Software Development :: Compilers",
+          "Topic :: Text Processing :: Linguistic",
+        ]
+
+        requires-python = ">=3.8"
+        license.text = "MIT"
+        readme = "README.md"
+
+        [project.optional-dependencies]
+        core = ["tree-sitter~=0.21"]
+
+        [tool.cibuildwheel]
+        build = "cp38-*"
+        build-frontend = "build"
+      '')
+      (writeTextDir "tests/test_language.py" ''
+        from ${snakeCaseName} import language
+        from tree_sitter import Language, Parser
+
+
+        def test_language():
+          lang = Language(language())
+          assert lang is not None
+          parser = Parser()
+          parser.language = lang
+          tree = parser.parse(bytes("", "utf-8"))
+          assert tree is not None
+      '')
+    ];
+  };
+
+  preCheck = ''
+    # https://github.com/NixOS/nixpkgs/issues/255262
+    rm -r ${snakeCaseName}
```

Comment:
The pytestCheckHook uses the command python -m pytest, and as explained in [this comment](https://github.com/NixOS/nixpkgs/issues/255262), this command adds the current directory to the search path. As a result, we end up executing tests for source files that do not contain the required shared object.

Looking at all the links in issue #255262, there are several ways to address this problem. A few examples include:

    Removing the source files (this is the approach we use).
    Using an alternative test command (e.g., pytest).

Reply:
Yeah so "looking at all the links" doesn't scale for readers, but regardless I don't understand the approach: this looks like we're removing the code under test entirely, so what's the point of running the test? Am I missing something?

Reply:
Sorry, I didn't mean for you to go through all the links. I was just pointing out that the solutions I found seem to involve some hacky workarounds for the `pytestCheckHook` behavior.

While I don't think this is directly related to the PR at hand, **I want to better understand the issue so I can explain it more clearly.** I am investigating righ now.

At a high level, my current understanding, it that it is related to how `pytest` loads the tests. We want `pytest` to use the compiled version that contains the shared object, rather than the source version that is not yet ready to be used. And without removing this directory, `pytest` loads the sources.

Reply:
### Summary of the issue
The `pytest -m` command used by the `pytestCheckHook` (`pkgs/development/interpreters/python/hooks/pytest-check-hook.sh`) adds the current directory to `sys.path` (documented [here](https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html#invoking-pytest-versus-python-m-pytest)).

With this setup, the current directory is prioritized before the Python site-packages folder in `sys.path`. Both locations contain a directory named `${snakeCaseName}`. When Python tries to load the module, it prioritizes the current directory, which lacks the compiled binding, causing the test to fail. *The compiled binding resides in `$out/${python.sitePackages}` after the build.*

This issue primarily affects bindings since the source does not include the compiled object.

---

### Example: `tree-sitter-python` grammar (`snakeCaseName` = `tree_sitter_python`)

During testing, the `sys.path` includes the following critical paths (among others). The second and third paths both contain a `tree_sitter_python` directory, but the first path is where the tests reside:
```plaintext
'/build/python-tree-sitter-python-source/tests'
'/build/python-tree-sitter-python-source'
'/nix/store/qs13k1z45jxh1q3hsda470f9r3i0pc86-python3.12-python-tree-sitter-python-0.24.3/lib/python3.12/site-packages'
...
```

---

#### Directory contents:

1. **First directory: `/build/python-tree-sitter-python-source`**
   ```
   .
   ├── build
   │   ├── lib.linux-x86_64-cpython-312
   │   │   └── tree_sitter_python
   │   │       ├── __init__.py
   │   │       ├── _binding.abi3.so
   │   │       └── binding.c
   ...
   ├── tests
   │   └── test_language.py
   ├── tree_sitter_python
   │   ├── __init__.py
   │   └── binding.c
   └── tree_sitter_python.egg-info
   ```

2. **Second directory: `$out/${python.sitePackages}`**
   ```
   ├── tree_sitter_python
   │   ├── __init__.py
   │   ├── _binding.abi3.so
   │   ├── binding.c
   │   └── __pycache__
   └── tree_sitter_python-0.24.3.dist-info
   ```

---

#### Failure mechanism:

When the test imports `tree_sitter_python`, Python looks in the first matching directory in `sys.path`. In this case, it prioritizes `/build/python-tree-sitter-python-source`, which does not contain the compiled `_binding.abi3.so` object. This results in an import error:
```
ImportError: No module named 'tree_sitter_python._binding'
```

---

I hope this raises the confusions.